File: clojure/example12.clj
; Example 12 - Graphical User Interfaces
; Author: Steven Yi <stevenyi@gmail.com>
; 2013.10.28
;
; This example demonstrates a slightly more advanced GUI example. 
; It uses a slider to allow setting the value of the frequency that
; the notes initiated by the button will play at.  
;
; Note: the actual use of update() here is not thread-safe.  In 
; real-world usage, we would need to drive Csound from a loop calling
; PerformKsmps to ensure thread-safety.  For this example, the updating
; generally works as there are few things demanding computation. 
; Also to note, using SetChannel on the Csound Object would be thread
; safe, as Csound internally protects channels with a spin lock. 


(import [javax.swing JFrame JButton JSlider])
(import [javax.swing.event ChangeListener])
(import [java.awt BorderLayout])
(import [java.awt.event ActionListener WindowAdapter])
(import [csnd6 csnd6 Csound CsoundPerformanceThread 
         CsoundMYFLTArray controlChannelType] 
        [java.util Random])


; this line turns off Csound's atexit handler as well as signal handlers
(csnd6/csoundInitialize 
  (bit-or csnd6/CSOUNDINIT_NO_ATEXIT 
          csnd6/CSOUNDINIT_NO_SIGNAL_HANDLER))


(def ctrl-chan-type 
  (bit-or (.swigValue controlChannelType/CSOUND_INPUT_CHANNEL) 
          (.swigValue controlChannelType/CSOUND_CONTROL_CHANNEL)))


(defn create-channel [csound channel-name]
  "Creates a Csound channel and returns a CsoundMYFLTArray wrapper object"
  (let [chn (CsoundMYFLTArray. 1)]
    (.GetChannelPtr csound (.GetPtr chn) channel-name ctrl-chan-type)
    chn))

; Our Orchestra for our project

(def orc "
sr=44100
ksmps=32
nchnls=2
0dbfs=1

gkpch chnexport \"freq\", 1

instr 1 
kpch port gkpch, 0.01, i(gkpch)
printk .05, gkpch
kenv linsegr 0, .05, 1, .05, .9, .8, 0
aout vco2 p4 * kenv, kpch
aout moogladder aout, 2000, .25 
outs aout, aout
endin")

(def c (Csound.)) ; create an instance of Csound

(doto c
  (.SetOption "-odac")  ; Set option for Csound 
  (.SetOption "-m7")    ; Set option for Csound 
  (.CompileOrc orc)     ; Compile Orchestra from String
  (.Start)              ; When compiling from strings, this call is necessary before doing any performing 
  )

(def freq-channel (create-channel c "freq"))

(def pt (CsoundPerformanceThread. c))
(.Play pt)

(let [frame (JFrame. "Csound API GUI Example")
      button (JButton. "Play Note")
      slider (JSlider. 80 600)
      ]
  (.SetValue freq-channel 0 (.getValue slider))
  (.addActionListener button 
                      (proxy [ActionListener] []
                        (actionPerformed [evt]
                          (.InputMessage pt "i1 0 2 .5"))))
  (.addWindowListener frame
                      (proxy [WindowAdapter] []
                        (windowClosing [evt]
                          (.Stop pt)
                          (.Join pt))))
  (.addChangeListener slider
                     (proxy [ChangeListener] []
                       (stateChanged [evt]
                         (.SetValue freq-channel 0 (.getValue slider)))))

  (.. frame getContentPane (add button BorderLayout/NORTH))
  (.. frame getContentPane (add slider))
  (.SetValue freq-channel 0 (.getValue slider)) 
  (doto frame
    (.setDefaultCloseOperation JFrame/EXIT_ON_CLOSE)
    (.pack)
    (.setVisible true)))

File: 6.x/java/src/csoundapiexamples/Example12.java
package csoundapiexamples;

import csnd6.Csound;
import csnd6.CsoundMYFLTArray;
import csnd6.CsoundPerformanceThread;
import csnd6.controlChannelType;
import csnd6.csnd6;
import java.awt.BorderLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JSlider;
import javax.swing.SwingUtilities;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

/**
 * Example 12 - Graphical User Interfaces Author: Steven Yi <stevenyi@gmail.com>
 * 2013.10.28
 *
 * This example demonstrates a slightly more advanced GUI example. It uses a
 * slider to allow setting the value of the frequency that the notes initiated
 * by the button will play at.
 *
 * Note: the actual use of SetValue() here is not thread-safe. In real-world
 * usage, we would need to drive Csound from a loop calling PerformKsmps to
 * ensure thread-safety. For this example, the updating generally works as there
 * are few things demanding computation. Also to note, using SetChannel on the
 * Csound Object would be thread safe, as Csound internally protects channels
 * with a spin lock.
 */
public class Example12 {

    /**
     * Creates a Csound Channel and returns a CsoundMYFLTArray wrapper object
     */
    public static CsoundMYFLTArray createChannel(Csound csound, String channelName) {
        CsoundMYFLTArray channel = new CsoundMYFLTArray(1);

        csound.GetChannelPtr(channel.GetPtr(), channelName,
                controlChannelType.CSOUND_CONTROL_CHANNEL.swigValue()
                | controlChannelType.CSOUND_INPUT_CHANNEL.swigValue());

        return channel;
    }

    public static void main(String[] args) {
        csnd6.csoundInitialize(
                csnd6.CSOUNDINIT_NO_ATEXIT | csnd6.CSOUNDINIT_NO_SIGNAL_HANDLER);

        // Defining our Csound ORC code within a String
        final String orc = "sr=44100\n"
                + "ksmps=32\n"
                + "nchnls=2\n"
                + "0dbfs=1\n"
                + "\n"
                + "gkpch chnexport \"freq\", 1\n"
                + "\n"
                + "instr 1 \n"
                + "kpch port gkpch, 0.05, i(gkpch)\n"
                + "printk .05, gkpch\n"
                + "kenv linsegr 0, .05, 1, .01, .9, .8, 0\n"
                + "aout vco2 p4 * kenv, kpch\n"
                + "aout moogladder aout, 2000, .25 \n"
                + "outs aout, aout\n"
                + "endin";

        SwingUtilities.invokeLater(new Runnable() {
            @Override
            public void run() {
                JFrame frame = new JFrame("Csound API GUI Example");
                JButton button = new JButton("Play Note");
                final JSlider slider = new JSlider(80, 600);
                
                final Csound c = new Csound();
                c.SetOption("-odac");
                c.SetOption("-m7");
                c.CompileOrc(orc);
                c.Start();

                final CsoundMYFLTArray freqChannel = createChannel(c, "freq");
                freqChannel.SetValue(0, slider.getValue());
                
                final CsoundPerformanceThread pt = new CsoundPerformanceThread(c);
                pt.Play();

                button.addActionListener(new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        pt.InputMessage("i1 0 2 .5");
                    }
                });

                slider.addChangeListener(new ChangeListener() {

                    @Override
                    public void stateChanged(ChangeEvent e) {
                        freqChannel.SetValue(0, slider.getValue());
                    }
                });

                frame.addWindowListener(new WindowAdapter() {
                    @Override
                    public void windowClosing(WindowEvent e) {
                        pt.Stop();
                        pt.Join();
                        c.Stop();
                    }
                });

                frame.getContentPane().add(button, BorderLayout.NORTH);
                frame.getContentPane().add(slider);
                frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
                frame.pack();
                frame.setVisible(true);
            }
        });

    }
}
